### PR TITLE
jewel: rbd: default features should be negotiated with the OSD

### PR DESCRIPTION
**Pull request by dillaman**
http://tracker.ceph.com/issues/19805

### CODE REVIEW

**Comment by dillaman** [comment]
@trociny I was thinking about this the other day and thinking of the OpenStack case, they unfortunately retrieve the "rbd_default_features" configuration property and then pass it down to the image create API. Therefore, attempting to create an image using a jewel-client against a pre-jewel OSD would still fail.

Would it be large of a hack / workaround to add a test to say if the supplied feature is the hard-coded default feature of the release (i.e. if 61 for >=jewel) it should also perform the negotiation with the OSD?

**Comment by trociny** [comment]
@dillaman This really looks like a hack to me and I would like we avoid if it is possible.
Do you know real cases for the OpenStack when they update clients before the server? I will try to learn more about Mirantis use cases tomorrow. According to Ceph official doc [1] the recommended way is to upgrade the server part fist.

May be fixing how the Cinder ensures layering support is enabled (as I understand this is a reason of such a behaviour) would be enough?

[1] http://docs.ceph.com/docs/master/install/upgrading-ceph/#upgrading-a-client

**Comment by dillaman** [comment]
@trociny Some distros (e.g. Xenial) ship Jewel client libraries but users might not be running a >=Jewel. puppet-ceph can now override the default features due to this issue [1] as a mitigation. I agree the issue should probably just be fixed in the various OpenStack projects.

[1] https://bugs.launchpad.net/puppet-ceph/+bug/1633552

**Comment by trociny** [comment]
@dillaman I am going to file a bug against the Cinder.

Do you think that changing the Cinder to call `create` with `features` parameter explicitly set only if `rbd_default_feature` does not contain layering could be a good solution?

**Comment by dillaman** [comment]
@trociny sounds good to me

**Comment by trociny** [comment]
Cinder bug report: https://bugs.launchpad.net/cinder/+bug/1689836

**Comment by smithfarm** [comment]
Jenkins re-test this please

**Comment by smithfarm** [comment]
The following three rbd-related PRs were included in a recent jewel integration branch. An rbd suite was run on said branch, and it failed rather spectacularly: http://pulpito.front.sepia.ceph.com/smithfarm-2017-08-21_20:07:45-rbd-wip-jewel-backports-distro-basic-smithi/

@dillaman Could you rebase/repush `wip-19805-jewel` to ceph-ci? I'll run an rbd suite on this single PR. Or do you have another suggestion?

File: src/tools/rbd/Utils.cc [hunk 1]
```diff
@@ -464,7 +470,9 @@ int get_image_options(const boost::program_options::variables_map &vm,
   }
 
   opts->set(RBD_IMAGE_OPTION_ORDER, order);
-  opts->set(RBD_IMAGE_OPTION_FEATURES, features);
+  if (features_set) {
+    opts->set(RBD_IMAGE_OPTION_FEATURES, features);
+  }
```
Comment (trociny) [inline src/tools/rbd/Utils.cc]: @dillaman  This makes `rbd clone` (when running without `--image-feature`) abort trying to get RBD_IMAGE_OPTION_FEATURES in `do_clone()` and failing:
```
tools/rbd/action/Clone.cc: 24: FAILED assert(r == 0)
```

**Comment by trociny** [comment]
@smithfarm Why you need wip-19805-jewel to be pushed to ceph-ci to run suite only on this PR?
You can fetch the branch to you local git with command:
```
git fetch origin pull/14977/head:wip-19873-jewel
```
(assuming `origin` is set to ceph/ceph).

Anyway, looking at failures, it seems they all are due to `rbd clone` command crash, and this is caused by this PR. I left a comment to Jason.

FYI, @dillaman might be not available until the next week.

**Comment by smithfarm** [comment]
@trociny As far as I know, Shaman only compiles `wip-...` branches if they appear in `ceph/ceph-ci.git` - that's why I pushed it there. (And teuthology will refuse to run on a branch that hasn't been built by Shaman.) Please tell me what I'm missing?

Anyway, thanks for tracking down the cause of the `rbd clone` assert failures.

**Comment by smithfarm** [comment]
@trociny Nevermind, now I get it :-) Yes, that's how I populate the integration branch.

**Comment by trociny** [comment]
@smithfarm 
> Please tell me what I'm missing?

I mean something like below:
```
git clone https://github.com/ceph/ceph.git
cd ceph
git fetch origin pull/14977/head:wip-19873-jewel
git remote add ci git@github.com:ceph/ceph-ci.git
git push ci wip-19873-jewel
```
Or merge it into your own branch instead of pushing `wip-19873-jewel` to `ci` in the last command:
```
git checkout your-wip-testing-branch
git merge wip-19873-jewel
git push ci your-wip-testing-branch
```

**Comment by dillaman** [comment]
update pushed

**Comment by smithfarm** [comment]
@dillaman @trociny Now we're getting a different test failure: http://pulpito.ceph.com/smithfarm-2017-08-24_15:58:59-rbd-wip-jewel-backports-distro-basic-smithi/

**Comment by smithfarm** [comment]
@dillaman This passed an rbd suite at http://tracker.ceph.com/issues/20613#note-22 and a rados suite at http://tracker.ceph.com/issues/20613#note-20

Please review.

**Pull request merged**